cachi2/core/package_managers/general.py

```diff
@@ -33,6 +44,172 @@ def download_binary_file(url, download_path, auth=None, insecure=False, chunk_si
+    :param str download_dir: Path to download file to
+    :raise FetchError: If download failed
+    """
+
+    # Parse git URL
+    # Example: git+ssh://git@github.com/kevva/is-positive.git#97edff6f52
+    parse_url = url.replace("git+ssh://", "https://").split("#")
+    repo_name = parse_url[0].removesuffix(".git").split("/")[-1]
+    repo_commit = parse_url[1] if len(parse_url) > 1 else None
+
+    try:
+        log.info(f"Clone repository - {parse_url[0]}")
+        with tempfile.TemporaryDirectory() as tmpdir:
+            with git.Repo.clone_from(parse_url[0], tmpdir) as repo:
+                repo.git.checkout(repo_commit)
+
+                log.info(f"Create tar archive for - {repo_name}")
+                with tarfile.open(os.path.join(download_dir, f"{repo_name}.tgz"), "w:gz") as tar:
+                    tar.add(tmpdir, arcname="package")
+
+    except Exception as exception:
+        log.error(f"Unsuccessful cloning or packing for {parse_url[0]}")
+        raise FetchError(
+            (
+                f"Could not process {parse_url[0]}. "
+                f"exception_name: {exception.__class__.__name__}, "
+                f"details: {exception}"
+            )
+        ) from None
+
+
+async def async_download_binary_file(
+    session: aiohttp_retry.RetryClient,
+    url: str,
+    download_dir: Path,
+    tarball_name: str,
+    auth: Optional[aiohttp.BasicAuth] = None,
+    chunk_size: int = 8192,
+):
+    """
+    Download a binary file (such as a TAR archive) from a URL using asyncio.
+
+    :param aiohttp_retry.RetryClient session: Aiohttp interface for making HTTP requests.
+    :param str url: URL for file download
+    :param str download_dir: Path to download file to
+    :param str tarball_name: Name of the file
+    :param aiohttp.BasicAuth auth: Authentication for the URL
+    :param int chunk_size: Chunk size param for Response.content.read()
+    :raise FetchError: If download failed
+    """
+    try:
+        log.debug(f"Download started - {tarball_name}")
+        async with session.get(url, auth=auth, raise_for_status=True) as resp:
+            with open(os.path.join(download_dir, tarball_name), "wb") as f:
+                while True:
+                    chunk = await resp.content.read(chunk_size)
+                    if not chunk:
+                        break
+                    f.write(chunk)
+
+    except Exception as exception:
+        log.error(f"Unsuccessful download: {tarball_name}")
+        # "from None" since we have the exception context in the logs
+        raise FetchError(
+            (
+                f"Could not download {tarball_name} from {url}. "
+                f"exception_name: {exception.__class__.__name__}, "
+                f"details: {exception}"
+            )
+        ) from None
+
+    log.debug(f"Download completed - {tarball_name}")
+
+
+async def get_dependencies(
+    download_dir: Path,
+    deps_to_download: List[str],
+    concurrency_limit: int,
+) -> List[str]:
```

Comment:
I don't think this function is general enough to be in `general.py`, it makes some npm-specific assumptions

But it would be very nice to have a general function for "download all these files asynchronously"

How about something like this:

```python
async def async_download_files(
    files_to_download: dict[str, Path],  # dict[url, download path]
    concurrency_limit: int,
) -> None:
```
The function shouldn't try to infer the download path from the url, the download path should be specified by the caller

The function also shouldn't git-clone anything (the caller shouldn't pass git urls)

Specifying the download paths and git-cloning git deps can be done by a `get_dependencies` function in `npm.py`

Reply:
I changed it, is it better now?

Reply:
Yup, it's actually general now :+1: 

(just the type hints are wrong)